File: rust/fabric/src/node/generic.rs
use crate::error::Result;
use crate::node::interface::{NodeConfig, NodeInterface};
use async_trait::async_trait;
use std::any::Any;

pub struct GenericNode {
    config: NodeConfig,
}

impl GenericNode {
    pub fn new(config: NodeConfig) -> Self {
        Self { config }
    }

    pub async fn set_config(&mut self, config: NodeConfig) {
        self.config = config;
    }
}

#[async_trait]
impl NodeInterface for GenericNode {
    fn get_config(&self) -> NodeConfig {
        self.config.clone()
    }

    async fn set_config(&mut self, config: NodeConfig) {
        self.config = config;
    }

    fn get_type(&self) -> String {
        "generic".to_string()
    }

    async fn handle_event(&mut self, _event: &str, _payload: &str) -> Result<()> {
        // Implement generic event handling logic here
        Ok(())
    }

    async fn update_config(&mut self, config: NodeConfig) {
        self.config = config;
    }

    fn as_any(&mut self) -> &mut dyn Any {
        self
    }
}

File: python/fabric/node/generic.py
import asyncio
from typing import Any, Dict
from .interface import NodeInterface, NodeConfig
from ..error import FabricError

class GenericNode(NodeInterface):
    def __init__(self, node_id: str, initial_config: Dict[str, Any]):
        self.node_id = node_id
        self.config = NodeConfig(node_id=node_id, config=initial_config)
        self.node_type = "generic"

    def get_config(self) -> NodeConfig:
        return self.config

    async def set_config(self, config: NodeConfig) -> None:
        self.config = config

    def get_type(self) -> str:
        return self.node_type

    async def handle_event(self, event: str, payload: str) -> None:
        print(f"Handling event: {event} with payload: {payload}")

    async def update_config(self, config: NodeConfig) -> None:
        await self.set_config(config)

    async def run(self, cancel_token: asyncio.Event) -> None:
        while not cancel_token.is_set():
            print(f"Generic node {self.node_id} running with config: {self.config}")
            await asyncio.sleep(1)